lib/combine_popolo_memberships.rb

```diff
@@ -1,30 +1,50 @@
 require 'combine_popolo_memberships/version'
+require 'date-range'
 
 # Takes multiple popolo membership arrays and combines them based on date.
 module CombinePopoloMemberships
-  def self.overlap(mem, term)
-    mS = mem[:start_date].to_s.empty? ? '0000-00-00' : mem[:start_date].to_s
-    mE = mem[:end_date].to_s.empty? ? '9999-99-99' : mem[:end_date].to_s
-    tS = term[:start_date].to_s.empty? ? '0000-00-00' : term[:start_date].to_s
-    tE = term[:end_date].to_s.empty? ? '9999-99-99' : term[:end_date].to_s
-
-    return unless mS < tE && mE > tS
-    (s, e) = [mS, mE, tS, tE].sort[1, 2]
-    {
-      _data: [mem, term],
-      start_date: s == '0000-00-00' ? nil : s,
-      end_date:   e == '9999-99-99' ? nil : e
-    }
+  class Membership
+    def initialize(raw)
+      @raw = raw
+    end
+
+    def start_date
+      @start_date ||= raw[:start_date].to_s.empty? ? '0000-00-00' : raw[:start_date].to_s
+    end
+
+    def end_date
+      @end_date ||= raw[:end_date].to_s.empty? ? '9999-99-99' : raw[:end_date].to_s
+    end
+
+    def date_range
+      @date_range ||= DateRange.new(start_date, end_date)
+    end
+
+    def overlap(other)
+      o = date_range.overlap(other.date_range)
+      return unless o
+      {}.merge(to_h).merge(other.to_h).tap do |h|
+        h[:start_date] = o.start_date
+        h[:end_date] = o.end_date
+      end
+    end
+
+    def to_h
+      raw
```

Comment:
Yeah, not sure what's best to do about this. For now I've just exposed `membership_hash` (the new name for `raw`) in the public interface and dropped this `#to_h` method. 7d39d37